### PR TITLE
[PREVIEW] OGM-983  native query improvement

### PR DESCRIPTION
**Pull request by schernolyas**
The PR contains improvements of support of native queries in Apache Ignite dialect

### CODE REVIEW

**Comment by DavideD** [comment]
Could you add a couple of details about what you are trying to improve in this PR?

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/impl/IgniteQueryDescriptor.java [hunk 1]
```diff
@@ -20,16 +20,19 @@
 
 	private final String sql;
 	private final List<Object> indexedParameters;
+	private final String table;
 	private final boolean hasScalar;
 //	private final List<Return> customQueryReturns;
 //	private final Set<String> querySpaces;
 
-	public IgniteQueryDescriptor(String sql, List<Object> indexedParameters, boolean hasScalar) {
+	public IgniteQueryDescriptor(String sql, String table, @Deprecated  List<Object> indexedParameters, boolean hasScalar) {
```
Comment (DavideD) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/impl/IgniteQueryDescriptor.java:28]: Why is that value deprecated?

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/query/nativequery/OscarWildePoem.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import javax.persistence.ElementCollection;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.NamedNativeQueries;
+import javax.persistence.NamedNativeQuery;
+
+@Entity
+//@todo fix problem with rename cache
+//@Table(name = OscarWildePoem.TABLE_NAME)
+
```
Comment (DavideD) [inline ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/query/nativequery/OscarWildePoem.java]: What is this comment about? You can simply remove the @Table if you need to

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/query/nativequery/OscarWildePoem.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+@NamedNativeQueries({
+		@NamedNativeQuery(name = "PortiaOscarWildePoem", query = "select _key,_val from OscarWildePoem  where  name='Portia' and author='Oscar Wilde'", resultClass = OscarWildePoem.class),
+		@NamedNativeQuery(name = "OscarWildePoemWithParameters", query = "select _key,_val from OscarWildePoem  where  name=?1 and author=?2", resultClass = OscarWildePoem.class)
+})
+public class OscarWildePoem {
+
+	//public static final String TABLE_NAME = "WILDE_POEM";
```
Comment (DavideD) [inline ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/query/nativequery/OscarWildePoem.java]: If it's not needed for the test, why don't you delete it?

File: ignite/src/test/resources/skipTests [hunk 4]
```diff
@@ -47,4 +47,7 @@ org.hibernate.ogm.backendtck.associations.collection.unidirectional.CollectionUn
 org.hibernate.ogm.backendtck.associations.manytoone.ManyToOneTest#testUnidirectionalManyToOne
 org.hibernate.ogm.backendtck.jpa.JPAResourceLocalTest#testBootstrapAndCRUD
 
+//wait answer from experts
+org.hibernate.ogm.backendtck.callbacks.PostUpdateTest
+
```
Comment (DavideD) [inline ignite/src/test/resources/skipTests]: PostUpdate should work now can you test it?

**Comment by DavideD** [comment]
I've added some comments, let me know

File: ignite/src/test/resources/skipTests (hunk 4 above)
Comment (schernolyas) [reply ignite/src/test/resources/skipTests]: Ok. I will retest after update core.

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/impl/IgniteQueryDescriptor.java (hunk 1 above)
Comment (schernolyas) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/impl/IgniteQueryDescriptor.java:28]: Main idea is migration to native hibernate query parameters.

File: ignite/src/test/resources/skipTests (hunk 4 above)
Comment (schernolyas) [reply ignite/src/test/resources/skipTests]: yes. it works. thanks! What did you do for it ? Please, open the secret :)

**Comment by schernolyas** [comment]
Main improvements are:
1) supports NamedNativeQueries for the dialect
2) support parametrized native queries

**Comment by schernolyas** [comment]
Main improvement showed in IgniteEntityManagerNativeQueryTest

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/query/nativequery/OscarWildePoem.java (hunk 3 above)
Comment (schernolyas) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/query/nativequery/OscarWildePoem.java]: hm .... I think I have to develop additional test for test support annotation @Table .  
Okey. I will clear comments and develop additional test.

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java [hunk 5]
```diff
@@ -621,12 +621,8 @@ else if ( !queryParameters.getPositionalParameters().isEmpty() ) {
 							backendQuery.getSingleEntityMetadataInformationOrNull().getEntityKeyMetadata()
 						);
 		}
-		else if ( backendQuery.getQuery().isHasScalar() ) {
+		else  {
 			throw new NotYetImplementedException();
-//			return new IgniteProjectionResultCursor( result, backendQuery.getQuery().getCustomQueryReturns(), queryParameters.getRowSelection() );
-		}
-		else {
-			throw new UnsupportedOperationException( "Not implemented yet" );
```
Comment (DavideD) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java]: @schernolyas You must be more precise when you write an error message,  if somebody receive this error he will wonder what's not yet implemented

Comment (DavideD) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java]: Oops...I confused the deletion with the remaining code

File: ignite/src/test/resources/skipTests (hunk 4 above)
Comment (DavideD) [reply ignite/src/test/resources/skipTests]: I just simplify the test, I don't remember exactly

**Comment by schernolyas** [comment]
Hi @DavideD !
Have you any comments for the PR? If no ... I will prepare PR that "ready for merge"

**Comment by schernolyas** [comment]
close till release 5.4

**Pull request closed without merging**